skimage/util/_flood_fill.py

```diff
@@ -0,0 +1,186 @@
+        `selem` is not None.
+    indices : bool, optional
+        If True, the output will be an array representing indices of local
+        maxima. If False, the output will be an array of 0's and 1's with the
+        same shape as `image`.
+    tolerance : ``selem`` type, optional
+        If None (default), adjacent values must be strictly equal to the
+        initial value of `image` at `seed_point`.  This is fastest.  If a value
+        is given, a comparison will be done at every point and this tolerance
+        on each side of the initial value will also be filled (inclusive).
+    inplace : bool, optional
+        If True, flood filling is applied to `image` inplace and nothing is
+        returned.  If False, the flood filled result is returned without
+        modifying the input `image`.  Ignored if `indices` is True.
+
+    Returns
+    -------
+    filled : ndarray or tuple[ndarray]
+        If `indices` is false, an array with the same shape as `image` is
+        returned with values equal to (or within tolerance) of the seed point
+        set to `new_value`.  If `indices` is true, a tuple of one-dimensional
+        arrays containing the coordinates (indices) of all found maxima.
+
+    See Also
+    --------
+    skimage.morphology.local_maxima
+    skimage.morphology.local_minima
+    skimage.morphology.h_maxima
+    skimage.morphology.h_minima
+
+    Notes
+    -----
+    The conceptual analogy of this operation is the 'paint bucket' tool in many
+    raster graphics programs.
+
+    Examples
+    --------
+    >>> from skimage.util import flood_fill
+    >>> image = np.zeros((4, 7), dtype=int)
+    >>> image[1:3, 1:3] = 1
+    >>> image[3, 0] = 1
+    >>> image[1:3, 4:6] = 2
+    >>> image[3, 6] = 3
+    >>> image
+    array([[0, 0, 0, 0, 0, 0, 0],
+           [0, 1, 1, 0, 2, 2, 0],
+           [0, 1, 1, 0, 2, 2, 0],
+           [1, 0, 0, 0, 0, 0, 3]])
+
+    Fill connected ones with 5, with full connectivity (diagonals included):
+
+    >>> flood_fill(image, (1, 1), 5)
+    array([[0, 0, 0, 0, 0, 0, 0],
+           [0, 5, 5, 0, 2, 2, 0],
+           [0, 5, 5, 0, 2, 2, 0],
+           [5, 0, 0, 0, 0, 0, 3]])
+
+    Fill connected ones with 5, with only cardinal direction connectivity:
+
+    >>> flood_fill(image, (1, 1), 5, connectivity=1)
+    array([[0, 0, 0, 0, 0, 0, 0],
+           [0, 5, 5, 0, 2, 2, 0],
+           [0, 5, 5, 0, 2, 2, 0],
+           [1, 0, 0, 0, 0, 0, 3]])
+
+    Fill with a tolerance:
+
+    >>> flood_fill(image, (0, 0), 5, tolerance=1)
+    array([[5, 5, 5, 5, 5, 5, 5],
+           [5, 5, 5, 5, 2, 2, 5],
+           [5, 5, 5, 5, 2, 2, 5],
+           [5, 5, 5, 5, 5, 5, 3]])
+    """
+    from ..morphology.extrema import (_resolve_neighborhood,
+                                      _set_edge_values_inplace, _fast_pad)
+    from ..morphology.watershed import _offsets_to_raveled_neighbors
+    from ._flood_fill_cy import _flood_fill_equal, _flood_fill_tolerance
```

Comment:
This gives me the idea that this function would fit better into the `morphology` module. It's related to `watershed` and deals with topology inside a digital image... which sounds like an morphological operation to me. 
Furthermore we'd gain the advantage that we could move these lazy imports outside the function.

I hope I haven't missed a discussion on this. Are their reasons to include the algorithm here? Are you using lazy imports for performance reasons?

Reply:
This decision was mine.  I view `flood_fill` as a very general operation, for example it can be used as a pseudo-threshold (with a wide tolerance anyway, other similar algorithms are in `filters` & `segmentation`), or for interactive drawing (`draw`, or `viewer`), color substitution (`color`), and yes also `morphology` - though I would posit that most users would think of it less as a morphological tool than the other uses.  With so many use cases, it seemed most appropriate to expose as a general utility function in `util` for ease of accessibility.  As the structure was based on extrema finding, I borrowed several helper functions.

The lazy imports are unfortunately required because the util module is brought in as a partial import during the build process, prior to Cython extensions being built.  Attempting to import any Cython at top level makes it impossible to build scikit-image.

Reply:
Reading this I have to agree, using utils seems justified. But maybe we could do something to enhance its discoverability by adding this function to the see-also section for similar functions (I only know of watershed)?
And thank you for providing your thoughts on this decision.

> The lazy imports are unfortunately required because the util module is brought in as a partial import during the build process, prior to Cython extensions being built. Attempting to import any Cython at top level makes it impossible to build scikit-image.

Interesting, I didn't know that!